### DISCUSSION TITLE
datacollector: enable getting nested attributes

### DISCUSSION THREAD

**Discussion by cylin928** [mesa/mesa#1715] [Ideas]
I try to have a model_reporters to access nested attributes like {"Col1": "a.b.c.d"}.
The current reporter will return None as the default for getattr().
One solution I found is to substitute getattr() at line 175 of the datacollection module to rgetattr() shown below.
```
import functools
def rgetattr(obj, attr, *args):
    def _getattr(obj, attr):
        return getattr(obj, attr, *args)
    return functools.reduce(_getattr, [obj] + attr.split('.'))
```
This will do the trick. Source: https://stackoverflow.com/questions/31174295/getattr-and-setattr-on-nested-subobjects-chained-properties

New code for collect():
```
def collect(self, model):
        """Collect all the data for the given model object."""
        def rgetattr(obj, attr, *args):
            def _getattr(obj, attr):
                return getattr(obj, attr, *args)
            return reduce(_getattr, [obj] + attr.split('.'))
        if self.model_reporters:
            for var, reporter in self.model_reporters.items():
                # Check if Lambda operator
                if isinstance(reporter, types.LambdaType):
                    self.model_vars[var].append(reporter(model))
                # Check if model attribute
                elif isinstance(reporter, str):
                    #self.model_vars[var].append(getattr(model, reporter, None))
                    self.model_vars[var].append(rgetattr(model, reporter, None))
                # Check if function with arguments
                elif isinstance(reporter, list):
                    self.model_vars[var].append(reporter[0](*reporter[1]))
                # TODO: Check if method of a class, as of now it is assumed
                # implicitly if the other checks fail.
                else:
                    self.model_vars[var].append(reporter())

        if self.agent_reporters:
            agent_records = self._record_agents(model)
            self._agent_records[model.schedule.steps] = list(agent_records)
```

**Comment by rht** [comment]
I'm not sure if the use case is common to the point that this should be in the library. In the meantime, you can define `rgetattr` in the model file, and do `{"Col1": rgetattr("a.b.c.d")}`.

**Reply by cylin928** [reply]
Sounds good.
The code will look like {"Col1": [rgetattr, [self, "a.b.c.d"]]}.
My code uses DotMap to incorporate sophisticated data and model structure. Therefore, I have a nested attribute structure. 
I agree with you that we don't know how often other people will appreciate the same feature.
Thank you for your prompt reply!
